File: pe-core/src/main/java/org/saar/core/model/InstancedArraysMesh.java
package org.saar.core.model;

import org.saar.lwjgl.opengl.constants.RenderMode;
import org.saar.lwjgl.opengl.objects.vaos.IVao;
import org.saar.lwjgl.opengl.utils.GlRendering;

public class InstancedArraysMesh extends MeshBase implements Mesh {

    private final RenderMode mode;
    private final int count;
    private final int instances;

    public InstancedArraysMesh(IVao vao, RenderMode mode, int count, int instances) {
        super(vao);
        this.mode = mode;
        this.count = count;
        this.instances = instances;
    }

    @Override
    protected void doDrawCall() {
        GlRendering.drawArraysInstanced(this.mode, 0, this.count, this.instances);
    }
}

File: pe-core/src/main/java/org/saar/core/mesh/InstancedArraysMesh.kt
package org.saar.core.mesh

import org.saar.lwjgl.opengl.constants.RenderMode
import org.saar.lwjgl.opengl.drawcall.InstancedArraysDrawCall
import org.saar.lwjgl.opengl.vao.IVao

class InstancedArraysMesh(
    private val vao: IVao,
    var drawCall: InstancedArraysDrawCall
) : Mesh {

    fun set(
        renderMode: RenderMode = this.drawCall.renderMode,
        first: Int = this.drawCall.first,
        count: Int = this.drawCall.count,
        instances: Int = this.drawCall.instances,
    ) {
        this.drawCall = InstancedArraysDrawCall(renderMode, first, count, instances)
    }

    override fun draw() {
        this.vao.bind()
        this.vao.enableAttributes()
        this.drawCall.doDrawCall()
    }

    override fun delete() {
        this.vao.delete()
    }
}